### PR TITLE
[build] Build checker generated files using a make rule.

### PR DESCRIPTION
**Pull request by ejgallego**
Currently, `configure.ml` does copy/link some files from `kernel` to
`checker` in an ad-hoc way. Instead, it is preferable to add a copy
rule to make and let it handle the dependencies properly.

### CODE REVIEW

**Comment by ejgallego** [comment]
I have no idea why this fails on windows but succeeds on Unix, very strange.

**Comment by ejgallego** [comment]
Ok, I think I've found the problem. Let's wait for CI but this should be ready now.

**Comment by ejgallego** [comment]
Ok @letouzey this is ready for review, and if positive, merge.

File: Makefile.checker
Comment (SkySkimmer) [inline Makefile.checker]: The change to this rule seems unnecessary.

Comment (ejgallego) [reply Makefile.checker]: Umm, what is your reasoning here?

**Note:** I am cheating, because I already tried without this rule and I already know that the sequential rule will fail.

Comment (SkySkimmer) [reply Makefile.checker]: ocamllibdep ignores the .ml arguments though.
so CHECKGENFILES should be an order-only dependency like OCAMLLIBDEP and shouldn't appear in the script part.

Comment (ejgallego) [reply Makefile.checker]: I see what you mean, that may be there for legacy reasons tho so if we want to do that cleanup I prefer to do it in a separate PR.

**Comment by ejgallego** [comment]
@letouzey ping

**Comment by ejgallego** [comment]
Requesting a review from the secondary maintainer.

**Comment by ejgallego** [comment]
I am undoing the assignment I did in the past [as in fact it happened when the rules were still written down and is non-conformant].

I am not sure if @letouzey is MIA, in that case I am requesting @maximedenes to take over.

**Comment by ejgallego** [comment]
More than 5 weeks opened, no reaction from anyone...

**Comment by ejgallego** [comment]
Anyway the bottom line here is that it took me maybe 15h to develop a feature (complete modularization of the codebase + dune support) and a order of magnitude more effort to have something happen, and still we are very far from anything going forward, despite most things being ready to ship.

So I can just imagine that any sane developer would want to find a project where contributing is a bit more efficient.

**Comment by maximedenes** [comment]
> You are the secondary maintainer tho, right?

Of the build system? No.

> So I can just imagine that any sane developer would want to find a project where contributing is a bit more efficient.

That's why I'm trying to address the problem

**Comment by ejgallego** [comment]
> Of the build system? No.

CODEOWNERS disagree [modulo the typo]

```
/Makefile*        @letouzey
# Secondary maintainer @maximdenes
```

**Comment by maximedenes** [comment]
Oh! I think that's a mistake.

**Comment by maximedenes** [comment]
I guess you or @silene could be secondary maintainers of the Makefile-based build system.

**Comment by ejgallego** [comment]
Well, would I be the secondary maintainer [which could make sense if we merge Dune], the problem we have here would still remain the same.

**Comment by maximedenes** [comment]
Yes, so I suggest: you become the secondary maintainer, and either we manage to get @letouzey look at build-related PRs, or we pick another primary maintainer for that component. Does it sound good?

**Comment by ejgallego** [comment]
> Yes, so I suggest: you become the secondary maintainer, and either we manage to get @letouzey look at build-related PRs, or we pick another primary maintainer for that component. Does it sound good?

I am OK being a maintainer of the build system, but note that my vision on it is that the current system has grown to be too complicated to maintain so I am favoring replacing some parts by domain-specific tools in the medium term.

A particular weakness I perceive in the current organization is that it doesn't make sense to me that the maintainers of `coq_makefile` and the main make system are disjoint. There is a clear overlap here, and as we discussed plugins and the library should be build by coq_makefile.

I am not sure tho it is worth doing the effort as Dune 1.0 has the capability to replace both the current make system and coq_makefile by a unified tool [coqdune, which is just a wrapper to coqdep has Dune doesn't understand the .vo dependencies yet]

So from my side I have a very clear roadmap on where I would like the build system to go, but yeah, it is my own roadmap and I guess I should present it at a WG. [I don't actually remember now if I have presented it yet?]

**Comment by maximedenes** [comment]
> So from my side I have a very clear roadmap on where I would like the build system to go, but yeah, it is my own roadmap and I guess I should present it at a WG. [I don't actually remember now if I have presented it yet?]

I don't remember seeing it, but I look forward to such a presentation.

**Comment by Zimmi48** [comment]
@letouzey is not in his office currently and he may be taking a few days off right now (as it is the holiday break - no more classes to give)

I can at least tell you what his general feeling about Dune is (although this is irrelevant for the present PR): "this can't become the main/only build system as long as Dune is not a standard package in Debian". Personally, I don't give a s**t about Debian, I'm just forwarding.

**Comment by maximedenes** [comment]
Weren't you complaining about our "non-standard" python packages recently, where by non-standard you meant that can be found on pypi but not Debian? :D

**Comment by ejgallego** [comment]
I'm sure Debían has all that is needed for building the documentation in the repositories, it is just that our build instructions are not very good yet.

Dune is a special case as it supports direct bootstrap using only ocaml.

**Comment by SkySkimmer** [comment]
I don't like the dynamic cp stuff, it seems likely to result in someone editing the checker files and losing their changes.
Maybe we should just have a `common/` directory with names, esubst, maybe univ/ugraph shared between coqtop and coqchk. The duplication doesn't give us anything AFAICT.

**Comment by ejgallego** [comment]
`cp` is usually ok if you consider this files to be "auto generated"; maybe we could mark them as read-only; a `kernel/lib` is not a bad option indeed, but it introduces yet another directory, also makes the `checker` folder not to be self-contained; I dunno.

Note that the `cp` already happens in windows, so it is not such a big difference, I could have the makefile to do a `ln` if on Unix; on the other hand, when this solution is used in `Dune`, the copied files only exist in the build tree, so you get less "trash". I dunno.

**Comment by ejgallego** [comment]
By the way, the reason this is needed is indeed due to the fact that the current configure method doesn't work if you are using a separate build directory.

**Comment by ejgallego** [comment]
I do find the "I may be editing the wrong file" argument valid but minor, indeed refactoring common parts of the kernel would be nice but on the other hand can be done in a separate PR.

File: Makefile.checker
Comment (maximedenes) [inline Makefile.checker]: How about adding a simple "DO NOT EDIT" header to the file, instead of just copying it? That's easy to do and would address @SkySkimmer 's concerns.

**Review by maximedenes — DISMISSED** [review]

File: Makefile.checker
Comment (ejgallego) [reply Makefile.checker]: Ok, warning added, not that is very visible but yeah; another choice would be to actually rename the files.

**Comment by ejgallego** [comment]
Anyone willing to self-assign this?

**Comment by Zimmi48** [comment]
If we could proceed and merge #7371, then @gares would be the secondary maintainer for `Makefile.checker`. But maybe we should actually do like in #7355 and #7403 and consider that the maintainers of `Makefile.checker` should be the maintainers of the checker itself (it would make much sense to me).

**Comment by ejgallego** [comment]
So the "order" based on ownership seems to be @letouzey @barras @gares @maximedenes , let's see what happens, but anyways I am going to comment in the Dune issue wrt the overall future of this project.

File: Makefile.checker
Comment (gares) [inline Makefile.checker]: IIRC `sed -i` is not ok on OSX (for sure not in busybox).
You could `( echo "DO NOT" ;  cat file1 ) > file2`

**Review by gares — CHANGES_REQUESTED** [review]
do no use `sed -i`

File: Makefile.checker
Comment (Zimmi48) [reply Makefile.checker]: It seems that it passed the MacOS tests though. But let's avoid breaking compatibility with BusyBox again if we can.

**Comment by Zimmi48** [comment]
I don't like either that we have to copy files between the kernel and the checker but this PR doesn't change anything to it. @ejgallego can you address @gares recent comment? Then, I think @gares will confirm that this is ready to merge.

File: Makefile.checker
Comment (ejgallego) [reply Makefile.checker]: That's strange, AFAICT `sed -i` is OK in all platforms, including Busybox https://busybox.net/downloads/BusyBox.html#sed

**Comment by ejgallego** [comment]
Yup the copy / link is not optimal, but this PR just moves it from `configure`-time to build time. I though making that hidden dependency explicit would be a good idea as for example the build system is unsound in windows today thanks to this.

As I commented above, copy is the most practical way when you want to have a separate build directory.

**Comment by Zimmi48** [comment]
If using `sed -i.bak` fixes the problem reported by @gares, then it's fine.

**Comment by Zimmi48** [comment]
Or I misunderstood. Where you saying that the current state of the patch was compatible?

**Comment by ejgallego** [comment]
Sorry, the push failed due to my flaky internet, should be good now. Indeed the previous patch not good for OSX's `sed`. This should work AFAICT. I think the CI uses GNU sed so that's why it passed.

**Comment by SkySkimmer** [comment]
BTW it seems we use sed -i in some Makefile.doc rule.

**Comment by ejgallego** [comment]
> BTW it seems we use sed -i in some Makefile.doc rule.

AFAICS it is a "safe" use.

**Review by gares — APPROVED** [review]

**Comment by Zimmi48** [comment]
CI passed. @gares Are you taking care of merging this?

**Pull request merged**
